src/System.Linq/src/System/Linq/Take.cs

```diff
@@ -96,5 +97,54 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
                 yield return element;
             }
         }
+
+        public static IEnumerable<TSource> TakeLast<TSource>(this IEnumerable<TSource> source, int count)
+        {
+            if (source == null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            if (count <= 0)
+            {
+                return EmptyPartition<TSource>.Instance;
+            }
+
+            return TakeLastIterator(source, count);
+        }
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(count > 0);
+
+            var queue = new Queue<TSource>();
+
+            using (IEnumerator<TSource> e = source.GetEnumerator())
+            {
+                while (e.MoveNext())
+                {
+                    if (queue.Count < count)
+                    {
+                        queue.Enqueue(e.Current);
+                    }
+                    else
+                    {
+                        do
+                        {
+                            queue.Dequeue();
+                            queue.Enqueue(e.Current);
+                        }
+                        while (e.MoveNext());
+                        break;
+                    }
+                }
+            }
+
+            while (queue.Count > 0)
+            {
+                yield return queue.Dequeue();
+            }
```

Comment:
@JonHanna I am trying to forego non-trivial optimizations in this PR because it should be a no-brainer to merge. I have already written an optimized version of `SkipLast` with a more lightweight circular buffer: https://github.com/dotnet/corefx/compare/master...jamesqo:optimized.skiptake.last But I am saving that for later.